data-service/src/main/java/com/amazonaws/blox/dataservice/repository/EnvironmentRepository.java

```diff
@@ -33,7 +34,8 @@ Environment createEnvironmentAndEnvironmentRevision(
   Environment getEnvironment(EnvironmentId environmentId)
       throws ResourceNotFoundException, InternalServiceException;
 
-  List<Environment> listEnvironments(Cluster cluster) throws InternalServiceException;
+  List<Environment> listEnvironments(Cluster cluster, Optional<String> environmentNamePrefix)
```

Comment:
The actual interesting code is following `@NonNull final Optional<String> environmentNamePrefix)`. You need to make sure the Optional is not null, :).    I would prefer to pass a nullable value to the method. And inside the method. you can use Optional.ofNullable to handle two different cases.

Reply:
No, it doesn't. I'm just used to use Optional to avoid null input to the method. Since you call this out, I will remove it.

Reply:
@wbingli well, the intention of use Optional is not checking whether the Optional is null or not. So if the Optional itself is null, a NPE is thrown. But I removed it anyway since you have concerns.

Reply:
Yeah, to be clear, using Optional in return types is totally okay, it's just the usage as method parameters (especially when the type itself is nullable) that's counter-indicated.

The exception to this is that certain service libraries will treat `Optional` parameters specially (e.g. the Swagger generator will make those `required=false` in the frontend model), so for those cases it may be useful.